ArduPlane/mode_takeoff.cpp

```diff
@@ -68,59 +68,78 @@ bool ModeTakeoff::_enter()
 
 void ModeTakeoff::update()
 {
+    // nothing works if we dont have a valid position!
+    if (!plane.current_loc.initialised()) {
+        return;
+    }
+    plane.prev_WP_loc = plane.current_loc;
+    const float dist = target_dist;
+    const float alt = target_alt;  
     if (!takeoff_started) {
-        // see if we will skip takeoff as already flying
+        const float direction = degrees(ahrs.yaw);
+        // if you are flying but had not used Mode Takeoff before
+        // then start_loc not set and use home altitude as start altitude
+        int32_t alt_cm;
+        if (start_loc.alt == 0 && plane.current_loc.get_alt_cm(Location::AltFrame::ABOVE_HOME, alt_cm)) {
+            start_loc.alt = plane.current_loc.alt - alt_cm;
+        }
+        // if already flying skip takeoff climb if above TKOFF_ALT, otherwise climb then loiter
+        // if not flying setup target location TKOFF_DIST from
+        // takeoff point, at a height of TKOFF_ALT
         if (plane.is_flying() && (millis() - plane.started_flying_ms > 10000U) && ahrs.groundspeed() > 3) {
-            gcs().send_text(MAV_SEVERITY_INFO, "Takeoff skipped - circling");
+            if (plane.current_loc.alt >= (start_loc.alt + alt*100.0)) {
```

Comment:
Should use the location methods to insure the frame is the same. EG:
```
float alt_diff;
if (plane.current_loc.get_alt_distance(start_loc, alt_diff) && (alt_diff > alt)) {
```

Reply:
why would the frame be different? these are locations and using their ASL alts

Reply:
They shouldn't be, but in general we do a terrible job of altitudes and keeping track of frames in plane. So if we can do it properly were less likely to mess it up in the future.